Abstract the state type for futexes

paulmenage opened the pull request:
In the same way that we expose `SmallAtomic` and `SmallPrimitive` to allow Windows to use a value other than an `AtomicU32` for its futex state, switch the primary futex state type from `AtomicU32` to `futex::Futex`.  The `futex::Futex` type should be usable as an atomic value with underlying primitive type equal to `futex::Primitive`. (`SmallAtomic` is also renamed to `SmallFutex`).

This allows supporting the futex API on systems where the underlying kernel futex implementation requires more user state than simply an `AtomicU32`.

All in-tree futex implementations simply define {`Futex`,`Primitive`} directly as {`AtomicU32`,`u32`}.

paulmenage commented:
A possible simplification that could make this change less intrusive (but also a little less obvious about exactly what is happening) would be to continue to require the underlying primitive to be a `u32` - so no need to add the `Primitive` type alias and we could use `std::sys::futex::AtomicU32` as the abstraction name, resulting in just a use `core::sync::atomic::AtomicU32` -> `use crate::sys::futex::AtomicU32` change in the users of the futex API.

Another option I looked at is to abstract via a `Futex` trait, which makes for a bigger change but results in (IMO) a clearer API definition between the futex implementation and futex users: https://github.com/rust-lang/rust/compare/master...paulmenage:rust:futex-abstraction-trait

joboet commented:
I'm fine with this change, but could you rename `Atomic` to `Futex`? That makes the intended purpose of the alias even clearer.

paulmenage commented:
Sure - should we also rename `SmallAtomic` to `SmallFutex`?

paulmenage commented:
OK, I updated it to do that. There was a name clash with the `Futex` marker trait in the Windows implementation, but it looks to me like you can remove that by having the futex functions be generic on `W: Waitable` and take an argument of `&W::Futex` to get the same type safety as before. Cc: @ChrisDenton since he added that trait originally.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11391374060/job/31695044644) [(plain)](https://github.com/rust-lang/rust/commit/471ff059fc9e784e93bb65efb58aa305581d608f/checks/31695044644/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.898 Building wheels for collected packages: reuse
#16 2.899   Building wheel for reuse (pyproject.toml): started
#16 3.146   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.147   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#16 3.148   Stored in directory: /tmp/pip-ephem-wheel-cache-ef0wk3cs/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.150 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.550 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.551 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
---
   |
13 | pub type Futex = AtomicU32;
   | --------------------------- previous definition of the type `Futex` here
...
22 | pub unsafe trait Futex {}
   | ^^^^^^^^^^^^^^^^^^^^^^ `Futex` redefined here
   = note: `Futex` must be defined only once in the type namespace of this module

error[E0404]: expected trait, found type alias `Futex`
  --> std/src/sys/pal/windows/futex.rs:32:25
  --> std/src/sys/pal/windows/futex.rs:32:25
   |
26 | / macro_rules! unsafe_waitable_int {
27 | |     ($(($int:ty, $atomic:ty)),*$(,)?) => {
29 | |             unsafe impl Waitable for $int {
...  |
...  |
32 | |             unsafe impl Futex for $atomic {}
33 | |         )*
34 | |     };
35 | | }
   | |_- in this expansion of `unsafe_waitable_int!`
   | |_- in this expansion of `unsafe_waitable_int!`
36 | / unsafe_waitable_int! {
37 | |     (bool, AtomicBool),
38 | |     (i8, AtomicI8),
39 | |     (i16, AtomicI16),
47 | |     (usize, AtomicUsize),
48 | | }
   | |_- in this macro invocation
   |
---

error[E0404]: expected trait, found type alias `Futex`
  --> std/src/sys/pal/windows/futex.rs:71:34
   |
71 | pub fn wake_by_address_single<T: Futex>(address: &T) {
   |
help: you might have meant to use `#![feature(trait_alias)]` instead of a `type` alias
   |
13 | pub trait Futex = AtomicU32;
13 | pub trait Futex = AtomicU32;
   |

error[E0404]: expected trait, found type alias `Futex`
  --> std/src/sys/pal/windows/futex.rs:78:31
   |
78 | pub fn wake_by_address_all<T: Futex>(address: &T) {
   |
help: you might have meant to use `#![feature(trait_alias)]` instead of a `type` alias
   |
13 | pub trait Futex = AtomicU32;
13 | pub trait Futex = AtomicU32;
   |

error[E0404]: expected trait, found type alias `Futex`
  --> std/src/sys/pal/windows/futex.rs:90:22
   |
90 | pub fn futex_wake<T: Futex>(futex: &T) -> bool {
   |
help: you might have meant to use `#![feature(trait_alias)]` instead of a `type` alias
   |
13 | pub trait Futex = AtomicU32;
13 | pub trait Futex = AtomicU32;
   |

error[E0404]: expected trait, found type alias `Futex`
  --> std/src/sys/pal/windows/futex.rs:95:26
   |
95 | pub fn futex_wake_all<T: Futex>(futex: &T) {
   |
help: you might have meant to use `#![feature(trait_alias)]` instead of a `type` alias
   |
13 | pub trait Futex = AtomicU32;
```

</details>

paulmenage commented:
> but it looks to me like you can remove that by having the futex functions be generic on W: Waitable and take an argument of &W::Futex to get the same type safety as before.

Ah, you could get the type safety but you don't get the type inference. I'll rename the old `Futex` trait on Windows.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11391506844/job/31695467878) [(plain)](https://github.com/rust-lang/rust/commit/c348b48496428c1ed7d3967e5cda81a6f1278e60/checks/31695467878/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.699 Building wheels for collected packages: reuse
#16 2.700   Building wheel for reuse (pyproject.toml): started
#16 2.949   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 2.950   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#16 2.951   Stored in directory: /tmp/pip-ephem-wheel-cache-8vwdus35/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 2.953 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.358 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.358 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.5s
```

</details>

paulmenage committed:
Abstract the state type for futexes

In the same way that we expose SmallAtomic and SmallPrimitive to allow
Windows to use a value other than an AtomicU32 for its futex state, this
patch switches the primary futex state type from AtomicU32 to
futex::Atomic.  The futex::Atomic type should be usable as an atomic
value with underlying primitive type equal to futex::Primitive.

This allows supporting the futex API on systems where the underlying
kernel futex implementation requires more state than simply an
AtomicU32.

All in-tree futex implementations simply define {Atomic,Primitive}
directly as {AtomicU32,u32}.

joboet commented:
Looks fine.
@bors r+

bors commented:
:pushpin: Commit cf7ff15a0ddfe0713dd794cd39eb3b3b58ba8d27 has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet cf7ff15a0ddfe0713dd794cd39eb3b3b58ba8d27 -->
<!-- homu: {"type":"Approved","sha":"cf7ff15a0ddfe0713dd794cd39eb3b3b58ba8d27","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

ChrisDenton commented:
Yep this looks fine to me too. Making the implementation more flexible makes sense.

(btw, sorry I've not been around as much lately and I'm still catching up with things)

bors commented:
:hourglass: Testing commit cf7ff15a0ddfe0713dd794cd39eb3b3b58ba8d27 with merge b0c2d2e5b00339014cba0a31f634af8c146458ed...
<!-- homu: {"type":"BuildStarted","head_sha":"cf7ff15a0ddfe0713dd794cd39eb3b3b58ba8d27","merge_sha":"b0c2d2e5b00339014cba0a31f634af8c146458ed"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11404905419/job/31745408374)
Approved by: joboet
Pushing b0c2d2e5b00339014cba0a31f634af8c146458ed to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"joboet","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11404905419/job/31745408374"},"merge_sha":"b0c2d2e5b00339014cba0a31f634af8c146458ed"} -->

rust-timer commented:
Finished benchmarking commit (b0c2d2e5b00339014cba0a31f634af8c146458ed): [comparison URL](https://perf.rust-lang.org/compare.html?start=1350eead10c46b9d3c2007fe0ea7892b7d7337ab&end=b0c2d2e5b00339014cba0a31f634af8c146458ed&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1350eead10c46b9d3c2007fe0ea7892b7d7337ab&end=b0c2d2e5b00339014cba0a31f634af8c146458ed&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1350eead10c46b9d3c2007fe0ea7892b7d7337ab&end=b0c2d2e5b00339014cba0a31f634af8c146458ed&stat=max-rss)
<details>
<summary>Results (secondary -2.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.6% | [-2.6%, -2.6%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1350eead10c46b9d3c2007fe0ea7892b7d7337ab&end=b0c2d2e5b00339014cba0a31f634af8c146458ed&stat=cycles:u)
<details>
<summary>Results (secondary -2.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.7% | [-2.7%, -2.7%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=1350eead10c46b9d3c2007fe0ea7892b7d7337ab&end=b0c2d2e5b00339014cba0a31f634af8c146458ed&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 782.585s -> 780.852s (-0.22%)
**Artifact size**: 333.84 MiB -> 333.83 MiB (-0.00%)
